### PR TITLE
XIN-117 add calcMasternodes, HookPenalty for v2, tests

### PR DESCRIPTION
**Pull request by wgr523**

### CODE REVIEW

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 1]
```diff
@@ -1328,3 +1334,50 @@ func (x *XDPoS_v2) GetCurrentEpochSwitchBlock(chain consensus.ChainReader, block
 	epochNum := x.config.XDPoSV2Block.Uint64()/x.config.Epoch + uint64(epochSwitchInfo.EpochSwitchBlockInfo.Round)/x.config.Epoch
 	return currentCheckpointNumber, epochNum, nil
 }
+
+func (x *XDPoS_v2) calcMasternodes(chain consensus.ChainReader, blockNum *big.Int, parentHash common.Hash) ([]common.Address, []common.Address, error) {
+	snap, err := x.getSnapshot(chain, blockNum.Uint64())
+	if err != nil {
+		log.Error("[calcMasternodes] Adaptor v2 getSnapshot has error", "err", err)
+		return []common.Address{}, []common.Address{}, err
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/engine.go]: return nil, nil, err is easier I think

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 2]
```diff
@@ -1328,3 +1334,50 @@ func (x *XDPoS_v2) GetCurrentEpochSwitchBlock(chain consensus.ChainReader, block
...
+	}
+	candidate := snap.NextEpochMasterNodes
+	if x.HookPenalty != nil {
+		penalties, err := x.HookPenalty(chain, blockNum, parentHash, candidate)
+		if err != nil {
+			log.Error("[calcMasternodes] Adaptor v2 HookPenalty has error", "err", err)
+			return []common.Address{}, []common.Address{}, err
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/engine.go]: same return nil, nil, err

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 3]
```diff
@@ -1328,3 +1334,50 @@ func (x *XDPoS_v2) GetCurrentEpochSwitchBlock(chain consensus.ChainReader, block
...
 	return currentCheckpointNumber, epochNum, nil
 }
+
+func (x *XDPoS_v2) calcMasternodes(chain consensus.ChainReader, blockNum *big.Int, parentHash common.Hash) ([]common.Address, []common.Address, error) {
+	snap, err := x.getSnapshot(chain, blockNum.Uint64())
+	if err != nil {
+		log.Error("[calcMasternodes] Adaptor v2 getSnapshot has error", "err", err)
+		return []common.Address{}, []common.Address{}, err
+	}
+	candidate := snap.NextEpochMasterNodes
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/engine.go]: `candidates` maybe

File: consensus/tests/test_helper.go [hunk 4]
```diff
@@ -442,8 +462,19 @@ func CreateBlock(blockchain *BlockChain, chainConfig *params.ChainConfig, starti
...
+			// Get last master node list from last v1 block
+			lastv1Block := blockchain.GetBlockByNumber(chainConfig.XDPoS.XDPoSV2Block.Uint64())
+			masternodesFromV1LastEpoch := decodeMasternodesFromHeaderExtra(lastv1Block.Header())
+			for _, v := range masternodesFromV1LastEpoch {
+				header.Validators = append(header.Validators, v[:]...)
+			}
+			// inject penalty for testing penalty hook if chainId is the specific one 1338
+			if chainConfig.ChainId.Int64() == 1338 {
+				header.Penalties = append(header.Penalties, signer[:]...)
+			}
```
Comment (liam-lai) [inline consensus/tests/test_helper.go]: either we have penalties in all test or we can insert this penalties header in specific test.

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 5]
```diff
@@ -1168,6 +1169,11 @@ func (x *XDPoS_v2) SetNewRoundFaker(newRound utils.Round, resetTimer bool) {
 	}
 	x.currentRound = newRound
 }
+func (x *XDPoS_v2) ProcessQCFaker(chain consensus.ChainReader, qc *utils.QuorumCert) error {
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/engine.go]: I think just call `ProcessQC` but add comment this function is for test only.

**Comment by liam-lai** [comment]
do we need to apply calcMasternodes on Yourturn function?

File: consensus/XDPoS/engines/engine_v2/engine.go (hunk 5 above)
Comment (wjrjerome) [reply consensus/XDPoS/engines/engine_v2/engine.go]: Why do we even need this function?

**Review by wjrjerome — APPROVED** [review]

File: consensus/XDPoS/engines/engine_v2/engine.go (hunk 5 above)
Comment (wgr523) [reply consensus/XDPoS/engines/engine_v2/engine.go]: For mining test, e.g., already have 899 and want to mine 900, we need a node to have QC of 899, round of 900, etc. And this func can prepare a node to have that

**Comment by wgr523** [comment]
calcMasternodes is already in Yourturn (one of the `if` branches)

**Review by liam-lai — APPROVED** [review]

**Pull request merged**